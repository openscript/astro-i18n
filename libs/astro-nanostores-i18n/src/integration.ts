import type { ComponentsJSON } from "@nanostores/i18n";
import type { AstroIntegration } from "astro";
import { addVirtualImports, createResolver } from "astro-integration-kit";
import { name } from "../package.json";

type Options = {
  translations?: Record<string, ComponentsJSON>;
};

/**
 * Astro integration for nanostores-i18n.
 *
 * This integration sets up the i18n configuration and provides the necessary
 * stores for managing translations and locale settings.
 *
 * @returns {AstroIntegration} The Astro integration object.
 */
const createPlugin = (options: Options): AstroIntegration => {
  const { resolve } = createResolver(import.meta.url);
  return {
    name,
    hooks: {
      "astro:config:setup": (params) => {
        const { config, logger } = params;

        if (!config.i18n) {
          logger.error(
            `The ${name} integration requires the i18n configuration in your Astro config. Please add it to your astro.config.ts file.`
          );
          return;
        }

        addVirtualImports(params, {
          name,
          imports: {
            [`${name}:runtime`]: `import { initializeI18n, useFormat, useI18n, currentLocale } from "${resolve("./runtime.js")}";

initializeI18n("${config.i18n.defaultLocale}", ${JSON.stringify(options.translations || {})});

export { useFormat, useI18n, currentLocale };
`,
          },
        });
      },
      "astro:config:done": (params) => {
        const { injectTypes } = params;
        injectTypes({
          filename: `${name}.d.ts`,
          content: `declare module "${name}:runtime" {
  import type { ComponentsJSON, Translations } from '@nanostores/i18n';
  export declare const currentLocale: import('nanostores').PreinitializedWritableAtom<string> & object;
  export declare const initializeI18n: (defaultLocale: string, translations: Record<string, Components>) => void;
  export declare const useFormat: () => import('@nanostores/i18n').Formatter;
  export declare const useI18n: <Body extends Translations>(componentName: string, baseTranslations: Body) => Body;
}
`,
        });
      },
    },
  };
};

export default createPlugin;
