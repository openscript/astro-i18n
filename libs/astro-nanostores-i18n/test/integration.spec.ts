import { describe, expect, it, vi } from "vitest";
import integration from "../src/integration";
import { AstroIntegrationLogger, BaseIntegrationHooks } from "astro";
import { addVirtualImports } from "astro-integration-kit";

vi.mock("astro-integration-kit", () => {
  return {
    addVirtualImports: vi.fn(),
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    createResolver: (_base: string) => ({
      resolve: (path: string) => [path].join("/"),
    }),
  };
});

describe("integration.ts", () => {
  it("should have a valid integration export", () => {
    expect(integration).toBeDefined();
    expect(typeof integration).toBe("function");
  });
  it("should log an error if i18n config is missing", () => {
    const mockParams = {
      config: {},
      logger: {
        error: (message: string) => {
          expect(message).toContain("The astro-nanostores-i18n integration requires the i18n configuration");
        },
      },
    };
    const i = integration({});
    const hook = i.hooks["astro:config:setup"];
    expect(hook).toBeDefined();
    if (hook) {
      hook(mockParams as Parameters<BaseIntegrationHooks["astro:config:setup"]>[0]);
    }
  });
  it("should register virtual module", () => {
    const mockParams = {
      config: { i18n: { defaultLocale: "en" } },
      logger: {} as AstroIntegrationLogger,
    };
    const i = integration({});
    const hook = i.hooks["astro:config:setup"];
    expect(hook).toBeDefined();
    if (hook) {
      hook(mockParams as Parameters<BaseIntegrationHooks["astro:config:setup"]>[0]);
    }
    expect(vi.mocked(addVirtualImports).mock.calls.length).toBe(1);
    expect(vi.mocked(addVirtualImports).mock.calls[0][1]).toMatchSnapshot();
  });
  it("should inject the types for the virtual module", () => {
    const mockParams = {
      injectTypes: (params: { filename: string; content: string }) => {
        expect(params.filename).toBe("astro-nanostores-i18n.d.ts");
        expect(params.content).toMatchSnapshot();
      },
    };
    const i = integration({});
    const hook = i.hooks["astro:config:done"];
    expect(hook).toBeDefined();
    if (hook) {
      hook(mockParams as Parameters<BaseIntegrationHooks["astro:config:done"]>[0]);
    }
  });
  it("should add middleware if configured", () => {
    const mockParams = {
      addMiddleware: vi.fn(),
      config: { i18n: { defaultLocale: "en" } },
    };
    const i = integration({ addMiddleware: true });
    const hook = i.hooks["astro:config:setup"];
    expect(hook).toBeDefined();
    if (hook) {
      hook(mockParams as Parameters<BaseIntegrationHooks["astro:config:setup"]>[0]);
    }
    expect(mockParams.addMiddleware).toHaveBeenCalledWith({
      entrypoint: "astro-nanostores-i18n/middleware",
      order: "pre",
    });
  });
});
