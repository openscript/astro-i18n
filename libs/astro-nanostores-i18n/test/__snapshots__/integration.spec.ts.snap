// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Integration Tests > should inject the types for the virtual module 1`] = `
"declare module "astro-nanostores-i18n:runtime" {
  import type { ComponentsJSON, Translations } from '@nanostores/i18n';
  export declare const currentLocale: import('nanostores').PreinitializedWritableAtom<string> & object;
  export declare const initializeI18n: (defaultLocale: string, translations: Record<string, Components>) => void;
  export declare const useFormat: () => import('@nanostores/i18n').Formatter;
  export declare const useI18n: <Body extends Translations>(componentName: string, baseTranslations: Body) => Body;
}
"
`;

exports[`Integration Tests > should register virtual module 1`] = `
{
  "imports": {
    "astro-nanostores-i18n:runtime": "import { initializeI18n, useFormat, useI18n, currentLocale } from "file:///workspaces/astro-loader-i18n/libs/astro-nanostores-i18n/src/integration.ts/./runtime.js";

initializeI18n("en", {});

export { useFormat, useI18n, currentLocale };
",
  },
  "name": "astro-nanostores-i18n",
}
`;
